Code/GraphMol/test1.cpp

```diff
@@ -1574,11 +1574,101 @@ void testGithub1843() {
   BOOST_LOG(rdErrorLog) << "Finished" << std::endl;
 }
 
+void testHasValenceViolation() {
+  BOOST_LOG(rdInfoLog) << " ----------> Testing Atom::hasValenceViolation()"
+                       << std::endl;
+
+  auto to_mol = [](const auto &smiles) {
+    int debugParse = 0;
+    bool sanitize = false;
+    auto mol = RDKit::SmilesToMol(smiles, debugParse, sanitize);
+    TEST_ASSERT(mol != nullptr);
+    mol->updatePropertyCache(false);
+    return mol;
+  };
+
+  //  All valid atoms
+  for (const auto &smiles : {
+           "C",
+           "C(C)(C)(C)C",
+           "S(C)(C)(C)(C)(C)C",
+           "O(C)C",
+           "[H]",
+           "[H+]",  // proton
+           "[H-]",
+           "[HH]",
+           "[He]",
+           "[C][C] |^5:0,1|",
+           "[H][Si] |^5:1|",
+           "[CH3+]",
+           "[CH3-]",
+           "[NH4+]",
+           "[Na]",
+           "[Na][H]",
+           "[Na]([H])[H]",
+           "[Og][Og]([Og])([Og])([Og])([Og])([Og])[Og]",
+           "[Lv-2]",
+           "[Lv-4]",
+           "[Lv+4]",
+           "[Lv+8]"
+           "*",              // dummy atom, which also accounts for wildcards
+           "*C |$_AP1;$|]",  // attachment point
+           "[*] |$_R1$|",    // rgroup
+       }) {
+    auto mol = to_mol(smiles);
+    for (auto atom : mol->atoms()) {
+      TEST_ASSERT(!atom->hasValenceViolation());
+    }
+  }
+
+  // First atom has a valence error!
+  for (const auto &smiles : {
+           // FIXME: commented out cases do not raise AtomValenceException
+           // when passing through the valence calculation code; will file
+           // RDKit issues to address within the valence code separately.
```

Comment:
Filed https://github.com/rdkit/rdkit/issues/6798 to address the commented out cases separately